pkg/proxy/healthcheck/healthcheck.go

```diff
@@ -0,0 +1,291 @@
+		}
+		if !entry.endpoints.Equal(*req.EndpointUids) {
+			// Compute differences (just for printing logs about additions and removals)
+			deletedEndpoints := entry.endpoints.Difference(*req.EndpointUids)
+			newEndpoints := req.EndpointUids.Difference(*entry.endpoints)
+			for _, e := range newEndpoints.List() {
+				glog.V(2).Infof("Adding local endpoint %s to LB health check for service %s", e, req.ServiceName.String())
+			}
+			for _, d := range deletedEndpoints.List() {
+				glog.V(2).Infof("Deleted endpoint %s from service %s LB health check (%d endpoints left)",
+					d, req.ServiceName.String(), len(*entry.endpoints))
+			}
+			// Replace the endpoints set with the latest
+			entry.endpoints = req.EndpointUids
+			h.serviceEndpointsMap[req.ServiceName] = entry
+		}
+	}
+}
+
+// handleServiceListenerRequest: receive requests to add/remove service health check listening ports
+func (h *ProxyHC) handleServiceListenerRequest(req *ProxyListenerRequest) bool {
+	sr, ok := h.serviceResponderMap[req.ServiceName]
+	if !req.Add {
+		if !ok {
+			return false
+		}
+		glog.Infof("Deleting HealthCheckListenPort for service %s port %d",
+			req.ServiceName, req.ListenPort)
+		delete(h.serviceResponderMap, req.ServiceName)
+		(*sr.listener).Close()
+		return true
+	} else if req.Add && ok {
+		// Responder for ServiceName already exists
+		if req.ListenPort == sr.listenPort {
+			return false
+		}
+		glog.Infof("HealthCheckListenPort for service %s changed from %d to %d - closing old listening port",
+			req.ServiceName, sr.listenPort, req.ListenPort)
+		delete(h.serviceResponderMap, req.ServiceName)
+		(*sr.listener).Close()
+	}
+	// Create a service responder object and start listening and serving on the provided port
+	glog.V(2).Infof("Adding health check listener for service %s on nodePort %d", req.ServiceName, req.ListenPort)
+	server := http.Server{Addr: fmt.Sprintf(":%d", req.ListenPort), Handler: h}
+	ln, err := net.Listen("tcp", server.Addr)
+	if err != nil {
+		glog.Warningf("FAILED to listen on address %s (%s)", server.Addr, err)
+		return false
+	}
+	h.serviceResponderMap[req.ServiceName] = serviceResponder{serviceName: req.ServiceName,
+		listenPort: req.ListenPort,
+		listener:   &ln,
+		server:     &server}
+	go func(serviceName types.NamespacedName, p uint16, ln net.Listener, server http.Server) {
+		// Anonymous goroutine to block on Serve for this listen port - Serve will exit when the listener is closed
+		glog.V(2).Infof("Goroutine blocking on serving health checks for %s on port %d", serviceName, p)
+		if err := server.Serve(ln); err != nil {
+			glog.Infof("Proxy HealthCheck listen socket %d for service %s closed (%s)", p, serviceName, err)
+		}
+	}(req.ServiceName, req.ListenPort, ln, server)
+	return true
+}
+
+func (h *ProxyHC) handlerLoop() {
+	for {
+		// use separate channels for mutations vs health check servicing for fairness
+		select {
+		case req := <-h.hcRequestChannel:
+			h.handleHealthCheckRequest(req)
```

Comment:
you're serializing all requests into a single buffered channel, and then popping requests off one by one. That feels inefficient? why not just call handleHealthCheckRequest directly from serveHTTP and lock around the service map. I get that you're trying to ensure some queuing, but I don't know if it's worthy it, especially since we don't want to start failing health checks due to load. 

The other option is to spin up a bunch of worker goroutines that are popping off this channel.